File: 0931-minimum-falling-path-sum/0931-minimum-falling-path-sum.java
class Solution {
    private int getMaxSum(int i,int j ,int m,int matrix[][],int dp[][]){
    // Base Case
        if(j < 0 || j >= m) return (int) Math.pow(10,9);
        if(i == 0) return matrix[0][j];
        if(dp[i][j] != -1) return dp[i][j];
        int up = matrix[i][j] + getMaxSum(i-1,j,m,matrix,dp);
        int leftDia = matrix[i][j] + getMaxSum(i-1,j-1,m,matrix,dp);
        int rightDia = matrix[i][j] + getMaxSum(i-1,j+1,m,matrix,dp);
        return dp[i][j] = Math.min(up,Math.min(leftDia,rightDia));
    }
   
    public int minFallingPathSum(int[][] matrix) {
        int n = matrix.length;
        int m = matrix[0].length;
        int maxi = Integer.MAX_VALUE;
        int dp[][] = new int[n][m];
        for(int row[]:dp){
            Arrays.fill(row,0);
        }
        for(int j = 0; j < m; j++){
            dp[0][j] = matrix[0][j];
            // int ans = getMaxSum(n-1,j,m,matrix,dp);
            // maxi = Math.min(maxi,ans);
        }
        for(int i = 1; i < n; i++) {
            for(int j = 0; j < m; j++) {
                int up = matrix[i][j] + dp[i-1][j];
                int leftDia = (int) 1e9;
                if(j-1 >= 0) {
                   leftDia =  matrix[i][j] + dp[i-1][j-1];   
                }
                
                int rightDia = (int) 1e9;
                if(j+1 < m) {
                   rightDia = matrix[i][j] + dp[i-1][j+1];   
                }
                dp[i][j] = Math.min(up,Math.min(leftDia,rightDia));
            }
        }
        maxi = dp[n-1][0];
        for(int j = 1; j < m; j++) {
            maxi = Math.min(dp[n-1][j],maxi);
        }
        return maxi;
    }
} 

File: 0931-minimum-falling-path-sum/0931-minimum-falling-path-sum.cpp
class Solution {
private:
    int f(int i,int j,vector<vector<int>>& matrix,vector<vector<int>>& dp) {
        if( j < 0  || j >= matrix[0].size()) return 1e9;
        if(i == 0) return matrix[i][j];
        if(dp[i][j] != -1) return dp[i][j];
        int strainght = matrix[i][j] + f(i-1,j,matrix,dp);
        int leftD = matrix[i][j] + f(i-1,j-1,matrix,dp);
        int rightD = matrix[i][j] + f(i-1,j+1,matrix,dp);
        return dp[i][j] = min(strainght,min(leftD,rightD));
    }
public:
    int minFallingPathSum(vector<vector<int>>& matrix) {
        int n = matrix.size();
        
        // vector<vector<int>> dp(n,vector<int> (n+1,0));
        vector<int> prev(n,0);
        vector<int> curr(n,0);
        for(int j = 0; j < n; j++) {
            prev[j] = matrix[0][j];
        }
        for(int i = 1;i < n; i++) {
            for(int j = 0; j < n; j++) {
                int strainght = matrix[i][j] + prev[j];
                int leftD = matrix[i][j];
                if(j - 1 >= 0) {
                  leftD +=  prev[j-1];    
                } else {
                    leftD += 1e9;
                }
                
                int rightD = matrix[i][j];
                if(j+1 < n) {
                    rightD += prev[j+1];
                } else {
                    rightD += 1e9;
                }
                curr[j] = min(strainght,min(leftD,rightD));
            }
            prev = curr;
        }
        
        int mini = 1e9;
        for(int j = 0; j < n; j++) {
            
            mini = min(mini,prev[j]);
        }
        return mini;
    }
};